File: src/main/java/com/abrastat/general/Status.java
package com.abrastat.general;

import static com.abrastat.general.Status.Volatility.*;
import static com.abrastat.general.Status.Target.*;

public enum Status {

    // VOLATILE = Removed by switching
    // NONVOLATILE = Persists through switching
    // Perish Song seems to be the only move which affects both users.
    // TODO define Baton Pass-able properties

    ATTRACT(VOLATILE, OPPONENT),
    BURN(NONVOLATILE, OPPONENT),
    CONFUSION(VOLATILE, OPPONENT),
    CURSE(VOLATILE, OPPONENT),
    DESTINYBOND(VOLATILE, SELF),
    DISABLE(VOLATILE, OPPONENT),
    ENCORE(VOLATILE,OPPONENT),
    ENDURE(VOLATILE, SELF),
    FATIGUE(VOLATILE, SELF),
    FLINCH(VOLATILE,OPPONENT),
    FORESIGHT(VOLATILE,OPPONENT),
    FREEZE(NONVOLATILE, OPPONENT),
    HEALTHY(NONVOLATILE, SELF),
    LEECHSEED(VOLATILE,OPPONENT),
    LOCKON(VOLATILE,OPPONENT),
    NIGHTMARE(VOLATILE,OPPONENT),
    PARALYSIS(NONVOLATILE, OPPONENT),
    PERISHSONG(VOLATILE, BOTH),
    POISON(NONVOLATILE, OPPONENT),
    PROTECT(VOLATILE, SELF),
    REST(NONVOLATILE, SELF),
    SAFEGUARD(VOLATILE, SELF),
    SLEEP(NONVOLATILE, OPPONENT),
    SUBSTITUTE(VOLATILE, SELF),
    TOXIC(NONVOLATILE, OPPONENT),
    TRANSFORM(VOLATILE, SELF),
    WRAP(VOLATILE, OPPONENT);



    private Volatility volatility;
    private Target target;

    Status(Volatility volatility, Target target)  {
        this.volatility = volatility;
        this.target = target;
    }

    enum Volatility {
        VOLATILE,
        NONVOLATILE
    }

    enum Target {
        SELF,
        OPPONENT,
        BOTH
    }

}

File: src/main/java/com/abrastat/general/Status.kt
package com.abrastat.general

enum class Status(val volatility: Volatility, private val target: Target) {
    // NONVOLATILE = Persists through switching
    BURN(Volatility.NONVOLATILE, Target.OPPONENT),
    FAINT(Volatility.NONVOLATILE, Target.SELF),
    FREEZE(Volatility.NONVOLATILE, Target.OPPONENT),
    HEALTHY(Volatility.NONVOLATILE, Target.SELF),
    PARALYSIS(Volatility.NONVOLATILE, Target.OPPONENT),
    POISON(Volatility.NONVOLATILE, Target.OPPONENT),
    SLEEP(Volatility.NONVOLATILE, Target.OPPONENT),
    TOXIC(Volatility.NONVOLATILE, Target.OPPONENT),

    // VOLATILE = Removed by switching
    ATTRACT(Volatility.VOLATILE, Target.OPPONENT),
    CONFUSION(Volatility.VOLATILE, Target.OPPONENT),
    CURSE(Volatility.VOLATILE, Target.OPPONENT),
    DESTINYBOND(Volatility.VOLATILE, Target.SELF),
    DISABLE(Volatility.VOLATILE, Target.OPPONENT),
    ENCORE(Volatility.VOLATILE, Target.OPPONENT),
    ENDURE(Volatility.VOLATILE, Target.SELF),
    FATIGUE(Volatility.VOLATILE, Target.SELF),
    FLINCH(Volatility.VOLATILE, Target.OPPONENT),
    FORESIGHT(Volatility.VOLATILE, Target.OPPONENT),
    HAZE(Volatility.VOLATILE, Target.OPPONENT),
    LEECHSEED(Volatility.VOLATILE, Target.OPPONENT),
    LIGHTSCREEN(Volatility.VOLATILE, Target.SELF),
    LOCKON(Volatility.VOLATILE, Target.OPPONENT),
    NIGHTMARE(Volatility.VOLATILE, Target.OPPONENT),
    PERISHSONG(Volatility.VOLATILE, Target.BOTH),
    PROTECT(Volatility.VOLATILE, Target.SELF),
    REFLECT(Volatility.VOLATILE, Target.SELF),
    SAFEGUARD(Volatility.VOLATILE, Target.SELF),
    SUBSTITUTE(Volatility.VOLATILE, Target.SELF),
    TRANSFORM(Volatility.VOLATILE, Target.SELF),
    WRAP(Volatility.VOLATILE, Target.OPPONENT)
    ;
    // Perish Song seems to be the only move which affects both users.
    // TODO define Baton Pass-able properties

    enum class Volatility {
        VOLATILE, NONVOLATILE
    }

    internal enum class Target {
        SELF, OPPONENT, BOTH
    }
}